RFC adding third party Ed25519, HPKE, and CBOR libs

JensenPaul opened the pull request.

gsnedders commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,38 @@
+We're proposing adding these libraries (or some very similar ones) to the
+tools/ directory:
+
+An Ed25519 Python implementation:
+https://github.com/pyca/ed25519/blob/main/ed25519.py
```
This appears to be CC0, which I know different companies have differing policies around, both in terms of usage and any potential contributions.

dadrian replied:
CC0 is broadly considered to be the legal equivalent to the colloquial definition of "public domain", so I don't expect that to be an issue, although perhaps @alex could weigh in.

Creative Commons (non-zero) is generally what triggers more paranoia across different companies. However, I am not sure what the policy is for Web Platform Tests.

alex replied:
I have _absolutely_ no context here, but:

That repo is a pure-Python, but somewhat optimized, implementation of ed25519. It was implemented by taking djb's original ed25519 Python code and optimizing it. djb's original implementation is listed as being licensed "in the public domain": https://ed25519.cr.yp.to/software.html. However, "public domain" is not a concept that exists in all jurisdictions, thus we licensed our modifications under the most permissive license that should work everywhere. For all practical purposes, I do believe anyone should be encumbered by this license.

All of that said, for 99.999% of use cases, you should get your ed25519 from pyca/cryptography, which is dual BSD-3-Clause/Apache-2 licensed. It will be much faster, etc. The only reason to use ed25519.py is if you have a use-case that absolutely requires it be pure python. Bootstrapping the python packaging ecosystem is the only real use case I'm aware of.

JensenPaul replied:
I believe the use case here, wptserve, is in fact pure-Python, so ed25519.py is in fact a reasonable match for our needs.

[CC0](https://creativecommons.org/publicdomain/zero/1.0/) licensing also sounds very permissive, and it looks like web-platform-tests already contains some CC0 licensed code: https://github.com/web-platform-tests/wpt/blob/master/images/wpt-pr-bot-logo/wpt-pr-bot-logo-darkblue.svg?short_path=d6712fa#L4

I think we've resolved all concerns here and so I'm going to resolve this thread.  Feel free to unresolve this comment if this does pose a problem.

alex replied:
The fact that wptserve is pure python doesn't really mean it needs to have only pure python dependencies. As the maintainer of ed25519.py, I think it'd be a mistake to use it unless you have immovable constraints like `pip` does.

But at any rate, you've got the licensing history, do as you please.

gsnedders replied:
We've certainly tried to avoid compiled dependencies — especially those that have more requirements to build above and beyond what CPython requires to build.

It's certainly the case that we care about being able to run WPT tests (and thus the server) on relatively unusual architectures/OSes (an obvious example is that some of the WebKit ports are still maintaining Linux/arm32 support, for which no cryptography wheels are available — and there's been reluctance to add all the dependencies to build cryptography into the image used — but also there are more extreme cases such as macOS on Apple Silicon prior to its public announcement); any time where we add a compiled dependency it potentially adds a whole load of "you need to install X, Y, and Z before you can install the Python dependencies".

Thus, in general, unless there's a very compelling reason, we've had a strong bias for pure-Python dependencies.

JensenPaul replied:
I resolved this thread, as the library in question has been removed from this RFC.

gsnedders commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,38 @@
+Users of these libraries may need to update them from time to time if new
+functionality or fixes are required. This is likely not a big risk.
+
+The HPKE library proposed may require inclusion of some other dependent
+libraries and uses deno to build into a single JavaScript file. Perhaps
+it's simplest to commit the single transpiled JavaScript file.
```
Some plan about what to do here should probably just be in the "Details" section above. Practically I presume the only sensible plan here is "commit the single built JS file"? One obvious risks here would be "this makes it hard to debug" (would there be a sourcemap somewhere? can we have a non-minified single JS file?), and those should probably be in the "Risks" section.

morlovich replied:
Non-minified is doable (and it involved changing command done by one of the build steps to remove --minified).

gsnedders commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,38 @@
+
+## Details
+
+We're proposing adding these libraries (or some very similar ones) to the
+tools/ directory:
```
JS stuff belongs in `/resources` (if it's expected to be maintained as core WPT infrastructure — which probably isn't true in this case, I'd expect these to be maintained by people working on related features), or in `/common`.

See also #46.

morlovich replied:
So having it in third_party/ under a specific subdir sounds great until it's needed in more places... which may mean it's the right way of doing it until it's actually needed in more places rather than speculated to be...?

JensenPaul replied:
I updated the RFC to follow the advice here: https://github.com/web-platform-tests/rfcs/issues/46#issuecomment-587707539

morlovich commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,38 @@
+
+The Ed25519 library is intended to be used by test code running on wptserve that may receive an Ed25519 private key and message to sign that message, or a public key and signature to verify that signature.
+
+An HPKE JavaScript implementation:
+https://github.com/dajiaji/hpke-js
```
If you're familiar with the tooling there, I would appreciate advice on how to check that. The generated file has a few other notices which are also MIT, but I am not confident that every dep pulled into a single-file build produced a notice.

JensenPaul replied:
Everything in the single file build is prefixed by a comment specifying what typescript file it was transpiled from.  I looked through them all and found that:

- 42 files are from https://github.com/dajiaji/hpke-js, which is MIT licensed.
- 21 files are from https://github.com/paulmillr/noble-ciphers, which is MIT licensed.
- 1 file is from https://github.com/denoland/dnt, which is MIT licensed.

Given how permissive the MIT license is and that there are multiple web-platform-test components with that license, I'm resolving this as I don't think it presents a issue.  I'll update my RFC to indicate the MIT licensing here.  Feel free to unresolve this comment if this does pose a problem.

JensenPaul commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,38 @@
+An HPKE JavaScript implementation:
+https://github.com/dajiaji/hpke-js
+
+A CBOR JavaScript implementation:
+https://github.com/paroga/cbor-js/blob/master/cbor.js
```
This is okay, right?  I see several MIT licensed things in WPT already:
https://github.com/web-platform-tests/wpt/blob/master/compression/third_party/pako/LICENSE
https://github.com/web-platform-tests/wpt/blob/master/tools/wave/www/lib/davidshimjs/LICENSE
https://github.com/web-platform-tests/wpt/blob/master/density-size-correction/third_party/piexif/LICENSE.txt
https://github.com/web-platform-tests/wpt/blob/master/docs/assets/gh-fork-ribbon.scss

JensenPaul replied:
Given how permissive the MIT license is and that there are multiple web-platform-test components with that license, I'm resolving this as I don't think it presents a issue.  I'll update my RFC to indicate the MIT licensing here.  Feel free to unresolve this comment if this does pose a problem.

JensenPaul commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,34 @@
+## Summary
+
+Several new web APIs require additional cryptography and CBOR libraries to
+properly test. These libraries are to support cryptography and the CBOR
```
I added a few examples.

Ms2ger commented on rfcs/additional_cypto_cbor_libs.md:
```diff
@@ -0,0 +1,49 @@
+
+The HPKE library, once built/transpiled into one JavaScript file, may be slightly harder to debug, but we'll disable minification when building.
+
+[0] https://github.com/WICG/turtledove/blob/main/FLEDGE.md#623-additional-bid-keys 
+[1] https://github.com/WICG/turtledove/blob/main/FLEDGE_Key_Value_Server_API.md#query-api-version-2
+[2] https://github.com/WICG/turtledove/blob/main/FLEDGE_browser_bidding_and_auction_API.md
+[3] https://github.com/web-platform-tests/rfcs/issues/46#issuecomment-587707539
```
Editorial: please use markdown links instead

Ms2ger commented:
If the idea is to skip the CC0 library for now, please update the RFC along those lines

JensenPaul commented:
> If the idea is to skip the CC0 library for now, please update the RFC along those lines

Based on comments [1](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731606298),[2](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731641860),[3](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1733502728) I hadn't planned on skipping the CC0 library.

I believe this RFC hasn't changed in 5 months and all comments are answered.

past commented:
> > If the idea is to skip the CC0 library for now, please update the RFC along those lines
> 
> Based on comments [1](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731606298),[2](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731641860),[3](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1733502728) I hadn't planned on skipping the CC0 library.
> 
> I believe this RFC hasn't changed in 5 months and all comments are answered.

Including a CC0-licensed library (ed25519.py) is not resolved and there is no consensus in the WPT core team yet about it. Individual organizations need to consult their legal teams about it, which is going to take a while.

JensenPaul commented:
> > > If the idea is to skip the CC0 library for now, please update the RFC along those lines
> > 
> > 
> > Based on comments [1](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731606298),[2](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1731641860),[3](https://github.com/web-platform-tests/rfcs/pull/208#discussion_r1733502728) I hadn't planned on skipping the CC0 library.
> > I believe this RFC hasn't changed in 5 months and all comments are answered.
> 
> Including a CC0-licensed library (ed25519.py) is not resolved and there is no consensus in the WPT core team yet about it. Individual organizations need to consult their legal teams about it, which is going to take a while.

I've removed the CC0-licensed library from this RFC and added an MIT-licensed Ed25519 library to this RFC.
